File: cpp/64_Minimum_Path_Sum/64_Minimum_Path_Sum.cpp
#include<vector>
#include<iostream>
#include<algorithm>

using namespace std;

class Solution {
public:
	int minPathSum(vector<vector<int>>& grid) {
		int nr = grid.size();
		if (nr == 0) return 0;
		int nc = grid[0].size();

		vector<int> dp(nc + 1, INT_MAX); // 由于dp[r][c] = min(dp[r-1][c], dp[r][c-1]) + grid[r][c]，因此可以用一维数组节约空间
		dp[1] = 0;
		for (int r = 1; r < nr + 1; r++)
		{
			for (int c = 1; c < nc + 1; c++)
			{
				dp[c] = min(dp[c], dp[c - 1]) + grid[r - 1][c - 1];	// dp[r][c] = min(dp[r-1][c], dp[r][c-1])
			}
		}
		return dp.back();

		// 一致的思路，但稍微容易理解一点
		// 另外，由于更详细的分支，减少了 nr+nc 次 min 比较，稍微快了一些
		//int nr = grid.size();
		//if (nr == 0) return 0;
		//int nc = grid[0].size();

		//vector<int> dp(nc + 1, 0);
		//for (int r = 0; r < nr; r++)
		//{
		//	for (int c = 0; c < nc; c++)
		//	{
		//		if (r == 0 && c == 0) dp[c] = 0;	// dp[0][0] = 0
		//		else if (r == 0) dp[c] = dp[c - 1];	// dp[r][c] = dp[r][c-1]
		//		else if (c == 0) dp[c] = dp[c];		// dp[r][c] = dp[r-1][c]
		//		else dp[c] = min(dp[c], dp[c - 1]);	// dp[r][c] = min(dp[r-1][c], dp[r][c-1])

		//		dp[c] += grid[r][c];
		//	}
		//}
		//return dp[nc - 1];
	}
};

File: python/64. Minimum Path Sum.py
class Solution:
    def minPathSum(self, grid) -> int:
        n_rows = len(grid)
        n_cols = len(grid[0])
        PathSum = [[0 for _ in range(n_cols)] for _ in range(n_rows)]

        for r in range(0, n_rows):
            for c in range(0, n_cols):
                if r==0 and c==0:
                    PathSum[r][c] = grid[r][c]
                elif r==0:
                    PathSum[r][c] = PathSum[r][c-1] + grid[r][c]
                elif c==0:
                    PathSum[r][c] = PathSum[r-1][c] + grid[r][c]
                else:
                    PathSum[r][c] = min(PathSum[r-1][c], PathSum[r][c-1]) + grid[r][c]

        return PathSum[n_rows-1][n_cols-1]



m = [
  [1,3,1],
  [1,5,1],
  [4,2,1]
]
ExpectOutput = 7

solu = Solution()  # 先对类初始化，才能进行调用
temp = solu.minPathSum(m)
if (temp == ExpectOutput):
    print('right')
else:
    print('wrong')
    print(temp)